### ISSUE TITLE
Add better support for universal packages

### ISSUE THREAD

**Issue by gr2m** [pikapkg/builders#90]
This is a follow up issue to https://github.com/gr2m/universal-user-agent/issues/23#issuecomment-534765805

I've adopted Pika across most [`@octokit/*` packages](https://github.com/octokit/). Some of the packages or its dependencies provide different code for browsers and Node, such as [`universal-user-agent`](https://github.com/gr2m/universal-user-agent). This causes problems for people who require a build step, e.g. for Lambda deployments or current GitHub Actions. There are workarounds on the users' side, but ideally I'd like to fix it with the libraries.

I'd currently suggest to create a new plugin: `@pika/plugin-build-esm`. It will work similar to the current [`@pika/plugin-build-web`](https://github.com/pikapkg/builders/tree/master/packages/plugin-build-web). It will also add a `"module"` key to the package.json. 

I first thought to create a `@pika/plugin-build-universal` plugin, but that would require to duplicate code from `@pika/plugin-build-web` and I think that can be avoided.

For that to work, the `@pika/plugin-build-web` would need to be changed:

- If the `"module"` key already exists, create a `"browser"` key instead.

I would also suggest the following changes

- in `@pika/plugin-build-web`, add support for the `"browser'` key, set it to true by default or don't make it optional at all. See https://github.com/pikapkg/builders/issues/84. I currently use [a fork](https://www.npmjs.com/package/@gr2m/pika-plugin-build-web) for that reason.
- in [`@pika/plugin-bundle-web`](https://github.com/pikapkg/builders/tree/master/packages/plugin-bundle-web#options), set the `"browser'` key to true by default, or remove the option altogether. I wouldn't know why people would _not_ want this to be set, if the plugin explicitly builds for browsers.

For `universal-user-agent`, my current `"@pika/pack" configuration [looks like this](https://github.com/gr2m/universal-user-agent/blob/d671405f5c3fc9922a7f61468d8ec694866825a4/package.json#L34-L46):

```json
  "@pika/pack": {
    "pipeline": [
      [
        "@pika/plugin-ts-standard-pkg"
      ],
      [
        "@pika/plugin-build-node"
      ],
      [
        "@gr2m/pika-plugin-build-web"
      ]
    ]
  },
```

When the above points would be addressed, the configuration would look like this:

```json
```json
  "@pika/pack": {
    "pipeline": [
      [
        "@pika/plugin-ts-standard-pkg"
      ],
      [
        "@pika/plugin-build-node"
      ],
      [
        "@pika/plugin-build-esm"
      ],
      [
        "@pika/plugin-build-web"
      ]
    ]
  },
```

I also heavily realy on Pika CDN to distribute the libraries for browsers, so the bundles made by pika.dev would need to prefer the alternative `"browser"` entrypoint over "module", too. I'm not sure if that is currently the case?

**Comment by FredKSchott** [comment]
Thanks for the thoughtful writeup, @gr2m. I've gone back and forth on this, and don't really have a good answer. Someone summarized it well in another thread, that we have 3 entrypoints trying to represent 4 different cjs/esm + web/node combinations, an impossible task.

To make things worse, different tools understand these entrypoints differently Pika & Rollup prefer "module" over "browser" for ESM/web due to issues analyzing legacy UMD in the "browser" entrypoint, while Webpack prefers "module" over "main" for the `"node"` target because... well I'm not sure why. "module" has no plans as any official Node entrypoint/support, so I'm not sure what the thinking is here. Most likely better tree-shaking, but clearly at the expense of this sort of issue. EVERYONE can't prefer module and still support universal packages.

So no matter what we do with Pika, this will be an issue for any package consumers using bundlers for Node.js that will need to be solved with config. Pika is going to continue to recommend "module" for the web entrypoint and "main" for Node.js, given that Node will sort this out in time (once Node goes ahead with "main" ESM support, Webpack will have to go back to preferring "main" over "module" in the "node" target environment).

But for users like you who want to experiment here, I'm happy to make the changes suggested above. Maybe even generalizing this type of support into a general plugin `"entrypoint"` config that lets the plugin user configure which entrypoints are set by that plugin.

I'll file an issue with the Webpack repo as well, since they may be interested in prioritizing this with the next major release to get ahead of Node's planned support.

**Comment by gr2m** [comment]
> but in practice it was already taken by legacy UMD + web distributions which are harder to analyze/tree-shake/bundle.

I think with Pika you already went 100% in with supporting ECMAScript modules only, and left commonJS behind, which I love. So I wouldn't expect pika to bother with legacy UMD distributions.

As an alternative, to avoid conflicts with other tools and services, maybe pika could support a new key, like `"pika-browser"` or `"pika-module"`, something like that?

Either way, shall I start working on the new `@pika/plugin-build-esm` plugin and the change to `@pika/plugin-build-web`, which could set `"browser"` if `"module"` is already set, or alternatively set a pika namespaced key as suggested above.

**Comment by FredKSchott** [comment]
Following up on this: after some off-thread discussions we're moving forward with supporting an explicit "browser:module" entrypoint (name TBD) in Pika tools/services (@pika/web, Pika CDN, etc) for any packages that need to ship Node.js code under "module". Most people shouldn't need this, but for the few that do it will like those packages continue to work with Pika.

In the plugin ecosystem, we're going to add a generalized "entrypoint" config to all relevant plugins to override the default entrypoint used. So for example, `["@pika/plugin-build-web"]` will continue to publish to the "module" entrypoint, but you can override this with `["@pika/plugin-build-web", {"entrypoint": "browser:module"}]`.

This should support the workflow described above, while keeping "module" as the default web entrypoint.

This issue is a tough one because there's no right answer (every tool assumes something different about "module"). Once Node releases ESM support this should resolve itself. Until then, we're stuck here :)

**Comment by gr2m** [comment]
Would it make sense to allow setting the `"entrypoint"` option to an array, so I can set it to `["browser:module", "browser"]`? 

Other than that, for universal ESM modules with different code for Node & Browser, we'd need another pika plugin for the node ESM build, right? Or could that be handled by one of the existing ones, too?

**Comment by FredKSchott** [comment]
Merged in via ac46b76fd5775fddefd180ca515fe7dc46d03b4f, with support for both `"entrypoint": string` & `"entrypoint": [string]`! Thanks for all the research, work, and time you've done on this @gr2m ❤️
